bus/a2bus: Add Uthernet II support Wiznet w5100

ksherlock opened the pull request:
a2RetroSystems Uthernet II / Wiznet W5100/W5100S emulation

Wiznet W5100

Used in: Uthernet II (Apple II), Spectranet (ZX Spectrum)

Based on:

W5100, W5100S, W5200, W5500 datasheets
WIZnet ioLibrary Driver (https://github.com/Wiznet/ioLibrary_Driver) https://docs.wiznet.io/Product/iEthernet/W5100/
Uthernet II User's and Programmer's Manual
Testing with Uthernet II and W5100S-EVB-Pico

High-level UDP/TCP/ICMP emulation is included, which attempts to match the quirks.

With the Uthernet II it's often only used as a raw ethernet device (ip65, marinetti, a2osx) but there is some software that uses the TCP stack, eg [A2stream](https://github.com/oliverschmidt/A2Stream)

    mame apple2ee -sl3 uthernet2  -flop1 A2Stream.dsk

cuavas commented on src/devices/machine/w5100.cpp:
```diff
@@ -0,0 +1,4320 @@
+*/
+
+
+static const int MAX_FRAME_SIZE = 1514;
+static uint8_t ETHERNET_BROADCAST[6] = { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff };
```
A global non-constant array is usually a bad sign.

cuavas commented on src/devices/machine/w5100.cpp:
```diff
@@ -0,0 +1,4320 @@
+	save_item(NAME(m_identification));
+	save_item(NAME(m_irq_state));
+
+	save_pointer(STRUCT_MEMBER(m_tcp, snd_una), 4);
+	save_pointer(STRUCT_MEMBER(m_tcp, snd_nxt), 4);
+	save_pointer(STRUCT_MEMBER(m_tcp, rcv_nxt), 4);
```
I can’t see how using `save_pointer` on structures like this is going to work properly.  It’s for saving indeterminate length arrays.  The count is the number of elements in the array.

cuavas commented on src/devices/machine/w5100.cpp:
```diff
@@ -0,0 +1,4320 @@
+	m_interrupt_timer = timer_alloc(FUNC(w5100_base_device::interrupt_timer), this);
+	for (int sn = 0; sn < 4 ; ++sn)
+	{
+		emu_timer *t;
+		t = timer_alloc(FUNC(w5100_base_device::delayed_ack_timer), this);
+		t->set_param(sn);
+		m_delayed_ack_timers[sn] = t;
+
+		t = timer_alloc(FUNC(w5100_base_device::keep_alive_timer), this);
+		t->set_param(sn);
+		m_keep_alive_timers[sn] = t;
```
What’s the purpose of the local `t`?  It just makes the code more error-prone because you’ve got more local mutable state.

cuavas commented on src/devices/machine/w5100.cpp:
```diff
@@ -0,0 +1,4320 @@
+
+uint8_t w5100_base_device::read(uint16_t offset)
+{
+	if (m_common_registers[0] & MR_IND)
+	{
+		switch(offset & 0x03)
+		{
+			case IDM_OR:
+				return m_common_registers[0];
+				break;
+			case IDM_AR0:
+				return m_idm >> 8;
+				break;
+			case IDM_AR1:
+				return m_idm & 0xff;
+				break;
+			case IDM_DR:
+				offset = m_idm;
+
+				if ((m_common_registers[0] & MR_AI) && !machine().side_effects_disabled())
+				{
+					m_idm++;
+					if (m_device_type == dev_type::W5100 && (m_idm & 0x1fff) == 0x0000)
+						m_idm -= 0x2000;
+				}
+				break;
+		}
+	}
```
`break` following `return` doesn’t make sense.

cuavas commented on src/devices/machine/w5100.cpp:
```diff
@@ -0,0 +1,4320 @@
+
+void w5100_base_device::write(uint16_t offset, uint8_t data)
+{
+	if (m_common_registers[0] & MR_IND)
+	{
+		switch(offset & 0x03)
+		{
+			case IDM_OR:
+				offset = 0;
+				break;
+			case IDM_AR0:
+				m_idm = (m_idm & 0x00ff) | (data << 8);
+				return;
+				break;
+			case IDM_AR1:
+				m_idm = (m_idm & 0xff00) | data;
+				return;
+				break;
+			case IDM_DR:
+				offset = m_idm;
+				if (m_common_registers[0] & MR_AI)
+				{
+					m_idm++;
+					if (m_device_type == dev_type::W5100 && (m_idm & 0x1fff) == 0x0000)
+						m_idm -= 0x2000;
+				}
+				break;
+		}
+	}
```
`break` following `return` doen’t make sense.

cuavas commented on src/devices/machine/w5100.cpp:
```diff
@@ -0,0 +1,4320 @@
+	}
+
+	offset &= 0x7fff;
+
+	return m_address_space->read_byte(offset);
```
Using an `address_space` directly is inefficient – you usually want to use an `address_space_specific` or `address_space_cache`.

cuavas commented on src/devices/machine/w5100.cpp:
```diff
@@ -0,0 +1,4320 @@
+	write_socket_register(0x0300 | (offset & 0xff), data);
+}
+
+void w5100_base_device::write_socket_register(offs_t offset, uint8_t data)
+{
+	const bool W5100S = m_device_type == dev_type::W5100S;
+
+	int sn = offset >> 8;
+	offset &= 0xff;
+
+	uint8_t *socket = m_socket_registers[sn];
+
+	switch(offset)
+	{
+		case Sn_MR:
+			socket[offset] = data;
+			if ((sn == 0) && (data & Sn_MR_MF) == Sn_MR_MF)
+				set_promisc(false);
+			break;
+
+		case Sn_PORT0:
+		case Sn_PORT1:
+		case Sn_PROTO:
+		case Sn_TOS:
+		case Sn_TTL:
+		case Sn_FRAGR0:
+		case Sn_FRAGR1:
+			/* these update immediately */
+			socket[offset] = data;
+			break;
```
If you’re using big switch statements inside your address space handlers, it usually means you have a problem.  You should be getting the memory system to do the heavy lifting for you and using on handler for each logical function.

cuavas commented on src/devices/machine/w5100.cpp:
```diff
@@ -0,0 +1,4320 @@
+
+	m_address_space = &space(AS_PROGRAM);
+
+	m_address_space->install_rom(0x0000, 0x00ff, 0x3f00, m_common_registers);
+	m_address_space->install_write_handler(0x0000, 0x00ff, 0, 0x3f00, 0, write8sm_delegate(*this, FUNC(w5100_base_device::write_common_register)));
+
+	if (m_device_type == dev_type::W5100S)
+	{
+		// m_address_space->install_rom(0x0400, 0x07ff, 0x3800, m_socket_registers);
+		m_address_space->install_read_handler(0x0400, 0x07ff, 0, 0x3800, 0, read8sm_delegate(*this, FUNC(w5100_base_device::read_socket_register)));
+		m_address_space->install_write_handler(0x0400, 0x07ff, 0, 0x3800, 0, write8sm_delegate(*this, FUNC(w5100_base_device::write_socket_register)));
+	}
+	else
+	{
+		// on the w5100, socket 3 is mirrored.
+		m_address_space->install_read_handler(0x0400, 0x04ff, 0x00ff, 0x3800, 0, read8sm_delegate(*this, FUNC(w5100_base_device::read_socket_register_3)));
+		m_address_space->install_write_handler(0x0400, 0x04ff, 0x00ff, 0x3800, 0, write8sm_delegate(*this, FUNC(w5100_base_device::write_socket_register_3)));
+
+		m_address_space->install_read_handler(0x0400, 0x07ff, 0, 0x0000, 0, read8sm_delegate(*this, FUNC(w5100_base_device::read_socket_register)));
```
Why aren’t you supplying an address map function when configuring the address space as this is all static and installed on start.

cuavas reviewed (changes requested):
Look, I haven’t gone through all the code yet, but there are already a bunch of concerning things, and it‘s going to need a fairly major rework which will result in it needing to be reviewed from scratch anyway.

Before we go any further, please refactor the code to at least address these points:
* Don’t use a “device type” member.  Provide implementations for differing functionality in the derived classes, and install the appropriate handlers, etc.
* Use address map functions to set up your internal address space rather than installing handlers in `device_start`.
* Either allocate your buffers dynamically on start or let the memory system do it for you (e.g. with `.ram()` in address map functions or using `memory_share_creator`).
* Use the memory system to dispatch to handlers for specific functions rather than using big switch statements inside your handlers.
* If you can, split up the big “registers” array into meaningful pieces.

ksherlock commented:
I've rewritten it to move all the registers into discrete fields and split the sockets into a separate device to simplify their registers and tcp/udp support. The w5100s support was also removed to keep things simple since it's not needed for the Uthernet II card.
